src/codegen/compile.ml

```diff
@@ -11205,107 +11007,122 @@ and compile_prim_invocation (env : E.t) ae p es at =
     IC.canister_version env
 
   | OtherPrim "crc32Hash", [e] ->
-    SR.UnboxedWord32 Type.Nat32,
+    SR.Vanilla,
     compile_exp_vanilla env ae e ^^
-    E.call_import env "rts" "compute_crc32"
+    E.call_import env "rts" "compute_crc32" ^^
+    G.i (Convert (Wasm_exts.Values.I64 I64Op.ExtendUI32)) ^^
+    TaggedSmallWord.msb_adjust Type.Nat32
 
   | OtherPrim "idlHash", [e] ->
     SR.Vanilla,
     E.trap_with env "idlHash only implemented in interpreter"
 
 
   | OtherPrim "popcnt8", [e] ->
-    SR.UnboxedWord32 Type.Nat8,
-    compile_exp_as env ae (SR.UnboxedWord32 Type.Nat8) e ^^
-    G.i (Unary (Wasm.Values.I32 I32Op.Popcnt)) ^^
+    SR.UnboxedWord64 Type.Nat8,
+    compile_exp_as env ae (SR.UnboxedWord64 Type.Nat8) e ^^
+    G.i (Unary (Wasm_exts.Values.I64 I64Op.Popcnt)) ^^
     TaggedSmallWord.msb_adjust Type.Nat8
   | OtherPrim "popcnt16", [e] ->
-    SR.UnboxedWord32 Type.Nat16,
-    compile_exp_as env ae (SR.UnboxedWord32 Type.Nat16) e ^^
-    G.i (Unary (Wasm.Values.I32 I32Op.Popcnt)) ^^
+    SR.UnboxedWord64 Type.Nat16,
+    compile_exp_as env ae (SR.UnboxedWord64 Type.Nat16) e ^^
+    G.i (Unary (Wasm_exts.Values.I64 I64Op.Popcnt)) ^^
     TaggedSmallWord.msb_adjust Type.Nat16
   | OtherPrim "popcntInt8", [e] ->
-    SR.UnboxedWord32 Type.Int8,
-    compile_exp_as env ae (SR.UnboxedWord32 Type.Int8) e ^^
-    G.i (Unary (Wasm.Values.I32 I32Op.Popcnt)) ^^
+    SR.UnboxedWord64 Type.Int8,
+    compile_exp_as env ae (SR.UnboxedWord64 Type.Int8) e ^^
+    compile_shrU_const (TaggedSmallWord.shift_of_type Type.Int8) ^^
+    G.i (Unary (Wasm_exts.Values.I64 I64Op.Popcnt)) ^^
     TaggedSmallWord.msb_adjust Type.Int8
   | OtherPrim "popcntInt16", [e] ->
-    SR.UnboxedWord32 Type.Int16,
-    compile_exp_as env ae (SR.UnboxedWord32 Type.Int16) e ^^
-    G.i (Unary (Wasm.Values.I32 I32Op.Popcnt)) ^^
+    SR.UnboxedWord64 Type.Int16,
+    compile_exp_as env ae (SR.UnboxedWord64 Type.Int16) e ^^
+    compile_shrU_const (TaggedSmallWord.shift_of_type Type.Int16) ^^
```

Comment:
Is the shrU necessary?

Reply:
It is redundant. Thanks for noticing. I removed it.